### DISCUSSION TITLE
Is there a way to lazy-load plugin just before loading another plugin?

### DISCUSSION THREAD

**Discussion by schardev** [wbthomason/packer.nvim#946] [Q&A] [answered]
I have something like this in my plugins file:
````lua
use({
    {
        "hrsh7th/nvim-cmp",
        event = "InsertEnter",
        config = function()
            require("configs.nvim-cmp")
        end,
    },
    { "hrsh7th/cmp-buffer", after = "nvim-cmp" },
})
````
Since I lazy-loaded `nvim-cmp`, I only want to load it's dependencies like `cmp-buffer` only when it's actually required.

Now, the problem with above code is that `nvim-cmp` won't register `cmp-buffer` source when lazy-loaded as it needs to be loaded __before__ `nvim-cmp`. Hence doing so with `after` key in packer plugin spec won't work.

So, is there a way to load a plugin (`cmp-buffer`) only when (and __before__) loading a plugin that requires it (`nvim-cmp`).

**Comment by mblarsen** [comment]
There is a `:PackerLoad` command that loads opt plugins. I don't know if there is an API version of that.
But perhaps you could set:

```
  use({ "hrsh7th/cmp-buffer", opt = true }) -- this makes it lazy
```

and 

```
use({
        "hrsh7th/nvim-cmp",
        event = "InsertEnter",
        config = function()
            require("configs.nvim-cmp")
            cmd([[ PackerLoad cmp-buffer ]]) -- this loads cmp-buffer 
        end,
    })
```

I do this for lazy loading themes.

**Comment by arsham** [comment]
Try this:
```lua
use({
  "hrsh7th/nvim-cmp",
  requires = {
    { "hrsh7th/cmp-buffer",   after = "nvim-cmp" },
    {
      "L3MON4D3/LuaSnip",
      requires = { "rafamadriz/friendly-snippets" },
      config   = function() require("settings.luasnip") end,
      event    = { "BufNewFile", "BufRead", "InsertEnter" },
    },
    { "saadparwaiz1/cmp_luasnip", after = "nvim-cmp" },
  },
  after  = { "LuaSnip", "nvim-treesitter" } ,
  wants  = { "LuaSnip" },
  config = function() require("settings.cmp") end,
  event  = { "BufNewFile", "BufRead", "InsertEnter" },
})

```

I have added more dependencies here as examples. Anything in the `requires` section will be loaded once any of the `events` are fired, and after the `nvim-cmp` is loaded.

**Comment by schardev** [comment, marked as answer]
Apologies for getting back to this a bit late. Unfortunately the solution by @mblarsen does not seem to work with `nvim-cmp` and to be honest I do not like the verbosity of the solution by @arsham (though I haven't checked if it works or not). Also I'm starting to think this is because of how `cmp` is structured as a similar issue was posted there a while ago (see https://github.com/hrsh7th/nvim-cmp/issues/65 for more info).
